### PR TITLE
refactor: refactor functions related to querying referrers

### PR DESCRIPTION
**Pull request by Wwwsylvia**
1. Use zero digest (`sha256:0000000000000000000000000000000000000000000000000000000000000000`) for pinging referrers API
2. Check `NAME_UNKNOWN` error code

Related to #271
Related to #354 
Related to https://github.com/opencontainers/distribution-spec/issues/359
Signed-off-by: Lixia (Sylvia) Lei <lixlei@microsoft.com>

### CODE REVIEW

File: registry/remote/repository.go [hunk 1]
```diff
@@ -1318,6 +1332,10 @@ func (r *Repository) isReferrersAPIAvailable(ctx context.Context, desc ocispec.D
 		r.SetReferrersCapability(true)
 		return true, nil
 	case http.StatusNotFound:
+		if err := errutil.ParseErrorResponse(resp); errcode.IsErrorCode(err, errcode.ErrorCodeNameUnknown) {
+			// repository not found
+			return false, fmt.Errorf("%s: %w", ref.Repository, errdef.ErrNotFound)
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:1337]: I'm wondering should we return error here.
`pingReferrersAPI()` is currently called by `Push()` and manifest `Delete()`.

1. For `Push()`, this is called AFTER pushing the manifest.
2. For `Delete()` this is called BEFORE deleting the manifest.

That is, the repository must be existing.

File: registry/remote/errcode/errors.go [hunk 2]
```diff
@@ -16,13 +16,30 @@ limitations under the License.
...
 import (
+	"errors"
 	"fmt"
 	"net/http"
 	"net/url"
 	"strings"
 	"unicode"
 )
 
+const (
```
Comment (qweeah) [inline registry/remote/errcode/errors.go:27]: should we attach the link of related distribution doc here?

File: registry/remote/repository.go [hunk 3]
```diff
@@ -392,20 +402,24 @@ func (r *Repository) Referrers(ctx context.Context, desc ocispec.Descriptor, art
...
-			r.SetReferrersCapability(false)
-			return r.referrersByTagSchema(ctx, desc, artifactType, fn)
+		if !errcode.IsErrorResponseStatus(err, http.StatusNotFound) {
+			return err
 		}
-		return err
+
+		if errcode.IsErrorCode(err, errcode.ErrorCodeNameUnknown) {
+			// The repository name is not found, no fallback
+			return fmt.Errorf("%s: %w", r.Reference.Repository, errdef.ErrNotFound)
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:412]: Should we return error in this case of just hide the error for users?

File: registry/remote/repository.go [hunk 4]
```diff
@@ -392,20 +402,24 @@ func (r *Repository) Referrers(ctx context.Context, desc ocispec.Descriptor, art
...
+		}
+		if state == referrersStateSupported {
+			// The repository is known to support Referrers API, no fallback.
+			return fmt.Errorf("referrers not found for %s: %w", r.Reference, errdef.ErrNotFound)
```
[the remark this answers is not in the document]
Comment (qweeah) [reply registry/remote/repository.go:416]: Why not returning the error directly?

File: registry/remote/errcode/errors.go (hunk 2 above)
Comment (Wwwsylvia) [reply registry/remote/errcode/errors.go:30]: added!

File: registry/remote/repository.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/repository.go]: Updated

File: registry/remote/errcode/errors.go [hunk 5]
```diff
@@ -101,3 +118,16 @@ func (err *ErrorResponse) Unwrap() error {
...
+func IsErrorResponseStatus(err error, statusCode int) bool {
+	var errResp *ErrorResponse
+	return errors.As(err, &errResp) && errResp.StatusCode == statusCode
+}
+
+// IsErrorCode returns true if err is an Error and its Code equals to code.
+func IsErrorCode(err error, code string) bool {
+	var ec Error
+	return errors.As(err, &ec) && ec.Code == code
+}
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply registry/remote/errcode/errors.go:142]: There is a similar case about  [os.IsExist(err error) bool](https://pkg.go.dev/os#IsExist) and [os.IsNotExist(err error) bool](https://pkg.go.dev/os#IsNotExist). The argument was that [os.IsExist(err) != !os.IsNotExist(err)](https://stefanxo.com/go-anti-patterns-os-isexisterr-os-isnotexisterr/).

[the remark this answers is not in the document]
Comment (shizhMSFT) [reply registry/remote/errcode/errors.go:142]: The system library suggests the use of `error.Is()`. Similarly, adopters should use `error.As()`.

File: registry/remote/repository.go [hunk 6]
```diff
@@ -399,13 +409,18 @@ func (r *Repository) Referrers(ctx context.Context, desc ocispec.Descriptor, art
 
 	// The referrers state is unknown.
 	if err != nil {
-		if errors.Is(err, errdef.ErrNotFound) {
-			// A 404 returned by Referrers API indicates that Referrers API is
-			// not supported. Fallback to referrers tag schema.
-			r.SetReferrersCapability(false)
-			return r.referrersByTagSchema(ctx, desc, artifactType, fn)
+		var errResp *errcode.ErrorResponse
+		if ok := errors.As(err, &errResp); !ok || errResp.StatusCode != http.StatusNotFound {
```
Comment (shizhMSFT) [inline registry/remote/repository.go:413]: According to the [doc](https://pkg.go.dev/errors#As), it is better to have
```suggestion
		if errors.As(err, &errResp) || errResp.StatusCode != http.StatusNotFound {
```

File: registry/remote/repository.go [hunk 7]
```diff
@@ -40,14 +40,20 @@ import (
...
 	"oras.land/oras-go/v2/registry/remote/internal/errutil"
 )
 
-// dockerContentDigestHeader - The Docker-Content-Digest header, if present on
-// the response, returns the canonical digest of the uploaded blob.
-// See https://docs.docker.com/registry/spec/api/#digest-header
-// See https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pull
-const dockerContentDigestHeader = "Docker-Content-Digest"
+const (
+	// dockerContentDigestHeader - The Docker-Content-Digest header, if present on
```
Comment (qweeah) [inline registry/remote/repository.go:48]: nit: exceeds 80-byte width

File: registry/remote/repository.go [hunk 8]
```diff
@@ -1288,18 +1300,21 @@ func (s *manifestStore) updateReferrersIndexForPush(ctx context.Context, desc, s
 	return s.repo.delete(ctx, oldIndexDesc, true)
 }
 
-// isReferrersAPIAvailable returns true if the Referrers API is available for r.
-func (r *Repository) isReferrersAPIAvailable(ctx context.Context, desc ocispec.Descriptor) (bool, error) {
+// pingReferrersAPI returns true if the Referrers API is available for r.
+func (r *Repository) pingReferrersAPI(ctx context.Context) (bool, error) {
+	// limit the rate of pinging referrers API
+	r.referrersPingLock.Lock()
+	defer r.referrersPingLock.Unlock()
```
Comment (shizhMSFT) [inline registry/remote/repository.go:1307]: Can we add the lock after the line 1314 in order to make it lock free in the most cases?

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: registry/remote/repository.go (hunk 8 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:1315]: Updated.

File: registry/remote/errcode/errors.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/errcode/errors.go:142]: Made it internal.

File: registry/remote/repository.go (hunk 7 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go]: fixed

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**